### PR TITLE
Buildbot embedded unittests

### PR DESCRIPTION
**Pull request by rbouqueau**

### CODE REVIEW

File: include/tests.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+#define custom_message_assert(expr, msg)                             \
+    do {                                                             \
+        if (expr) {                                                  \
+            checks_passed++;                                         \
+        } else {                                                     \
+            checks_failed++;                                         \
+        }                                                            \
+        ((expr)                                                      \
+         ? gf_fatal_assert(0)                                        \
+         : printf("Assertion failed: %s\nMessage: %s\n\tFile: %s\nLine: %d\nFunction: %s\n", #expr, msg, __FILE__, __LINE__, __ASSERT_FUNCTION)); \
```
[the remark this answers is not in the document]
Comment (rbouqueau) [reply include/tests.h:23]: I propose to remove all the elements in that section that are unused. @soheibthriber Let me know if you disagree.

File: unittests/README [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Unit tests in GPAC
```
[the remark this answers is not in the document]
Comment (rbouqueau) [reply unittests/README:1]: Will do, thanks!

File: include/tests.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            if (verbose_ut) printf("Assertion passed: \"%s\", File: \"%s\", Line: %d, Function: \"%s\"\n", #expr, __FILE__, __LINE__, __ASSERT_FUNCTION); \
+            checks_passed++;                                         \
+        } else {                                                     \
+            checks_failed++;                                         \
+            gf_assert(0);                                            \
+        }                                                            \
+    } while (0)
+
+#define assert_false(expr) assert_true(!(expr))
+
```
[the remark this answers is not in the document]
Comment (rbouqueau) [reply include/tests.h:50]: I like them.

1. Maybe I would put them in MAJ to make it clear this belongs to the UT framework (and not GPAC in general).

2. Also we could limit to  ASSERT() and ASSERT_STR_EQ() I guess. Opinion?

File: include/tests.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    } while (0)
+
+#define assert_true(expr)                                            \
+    do {                                                             \
+        if (expr) {                                                  \
+            if (verbose_ut) printf("Assertion passed: \"%s\", File: \"%s\", Line: %d, Function: \"%s\"\n", #expr, __FILE__, __LINE__, __ASSERT_FUNCTION); \
+            checks_passed++;                                         \
+        } else {                                                     \
+            checks_failed++;                                         \
+            gf_assert(0);                                            \
```
[the remark this answers is not in the document]
Comment (rbouqueau) [reply include/tests.h:19]: No but using similar names is confusing, for sure :/

[the remark this answers is not in the document]
Comment (rbouqueau) [reply include/tests.h:19]: outdated comment

File: include/tests.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+#define custom_message_assert(expr, msg)                             \
+    do {                                                             \
+        if (expr) {                                                  \
+            checks_passed++;                                         \
+        } else {                                                     \
+            checks_failed++;                                         \
+        }                                                            \
+        ((expr)                                                      \
+         ? gf_fatal_assert(0)                                        \
```
[the remark this answers is not in the document]
Comment (rbouqueau) [reply include/tests.h]: outdated comment

File: include/tests.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <gpac/setup.h>
+
+#define unittest(suffix) void test_##suffix(void)
+
+extern int checks_passed;
+extern int checks_failed;
+
+static Bool verbose_ut = GF_FALSE;
```
[the remark this answers is not in the document]
Comment (rbouqueau) [reply include/tests.h:10]: Will do.

File: configure [hunk 7]
```diff
@@ -2167,6 +2171,8 @@ if test $cpu = "mips"; then
 fi
 echo ""
 echo "** GPAC $version rev$revision Core Configuration **"
+echo "Unit Tests: $unit_tests ('make unit_tests')"
+echo "#define GF_STATIC static" >> $TMPH
```
[the remark this answers is not in the document]
Comment (rbouqueau) [reply configure:2175]: Added in 672f66666. Should we also add a target in the msvc Solution ? xcode ? What's the usual way to add this?

File: unittests/build.sh [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  int test_##ut();                                                                                                     \
+  if(register_test(#ut, test_##ut)) {                                                                                  \
+    fprintf(stderr, "registration failed for unit test: %s\n", #ut);                                                   \
+    return EXIT_FAILURE;                                                                                               \
+  }
+
+int main(int argc, char *argv[])
+{'
+
+calls=$(find "$scriptDir/.." -path "*/unittests/*.c" | grep -v bin | xargs grep unittest | cut -d ":" -f 2)
```
[the remark this answers is not in the document]
Comment (rbouqueau) [reply unittests/build.sh]: It seems like readlink and realpath are non portable. I just put a 'cd' before 'find' in a8e7ddb19.

**Comment by rbouqueau** [comment]
Note from today's call:
- This is ok as a first UT runtime. Existing unittests can run on any runtime anyway.
- @aureliendavid to have a look at how to get rid of the double build.
- Remarks that  UTs should not leverage the ```configure``` script but a compile-time define (with -D) instead or with var env (as asan does).
- Failing with ```if (fatal_ut) gf_assert(0);``` doesn't display the report when failing.

**Comment by rbouqueau** [comment]
Possible additional discussed tests:
- ```gpac -h dash``` doesn't return ```dasher```.
- 48e63767142f55cb78605cb423c57edd514a8709

**Comment by rbouqueau** [comment]
@aureliendavid Are we ok now? https://buildbot.gpac.io/#/grid?branch=buildbot-embedded-unittests

**Comment by aureliendavid** [comment]
it seems ok on the https://github.com/gpac/gpac/tree/buildbot-embedded-unittests branch

since it's a different branch than the one in this PR, maybe create a new PR for the new branch with a reference to this PR so that we don't lose the discussion? (or you can just merge the main repo branch directly if you prefer)

**Comment by rbouqueau** [comment]
Both branches are now aligned. @jeanlf Please merge if ok for you!

**Pull request merged**